File: Capítulo_8/Sondagem linear/Hash Table Linear Probing/Método remove/remove.js
import { defaultToString, ValuePair } from "../../../util/util";

// Classe HashTableLinearProbing
class HashTableLinearProbing {
    constructor(toStrFn =  defaultToString) {
        this.toStrFn = toStrFn;
        this.table = {};
    }

    // Método loseloseHashCode
    loseloseHashCode(key) {
        if (typeof key === "number") {
            return key;
        }
        const tableKey = this.toStrFn(key);
        let hash = 0;
        for (let i = 0; i < tableKey.length; i++) {
            hash += tableKey.charCodeAt(i);
        }
        return hash % 37;
    }

    // Método hashCode
    hashCode(key) {
        return this.loseloseHashCode(key);
    }

    // Método remove
    remove(key) {
        const position = this.hashCode(key);
        if (this.table[position] != null) {
            if (this.table[position].key === key) {
                delete this.table[position];
                this.verifyRemoveSideEffect(key, position);
                return true;
            }
            let index = position + 1;
            while (this.table[index] != null && this.table[index].key !== key) {
                index++;
            }
            if (this.table[index] != null && this.table[index].key === key) {
                delete this.table[index];
                this.verifyRemoveSideEffect(key, index);
                return true;
            }
        }
        return false;
    }
}

File: Capítulo_8/Sondagem linear/Hash Table Linear Probing/Método remove/Código em Python/remove.py
# defaultToString:
def defaul_to_string(item):
    if item is None:
        return "NONE"
    elif isinstance(item, str):
        return f"{item}"
    return str(item)

# Classe ValuePair
class ValuePair:
    def __init__(self, key, value):
        self.key = key
        self.value = value

    def __str__(self):
        return f"[#{self.key}: {self.value}]"
    
# Classe HashTableLinearProbing
class HashTableLinearProbing:
    def __init__(self, to_str_fn = defaul_to_string):
        self.to_str_fn = to_str_fn
        self.table = {}

    # Método loseloseHashCode
    def loselose_hash_code(self, key):
        if isinstance(key, (int, float)):
            return key
        table_key = self.to_str_fn(key)
        hash = 0
        for i in range(len(table_key)):
            hash += ord(table_key[i])
        return hash % 37
    
    # Método hash_code
    def has_code(self, key):
        return self.loselose_hash_code(key)
    
    # Método remove
    def remove(self, key):
        position = self.hash_code(key)
        table = self.get_table()
        if position in table:
            if table.get(position).key == key:
                del table[position]
                print(self.verify_remove_side_effect(key, position))
                return True
            index = position + 1
            while index in table and table.get(index).key != key:
                index += 1
            if index in table and table.get(index).key == key:
                del table.get[index]
                self.verify_remove_side_effect(key, index)
                return True
        return False
